espnet2/tts/gst/style_encoder.py

```diff
@@ -0,0 +1,238 @@
+        conv_stride (int, optional):
+            Stride size of conv layers in the reference encoder.
+        gru_layers (int, optional): The number of GRU layers in the reference encoder.
+        gru_units (int, optional): The number of GRU units in the reference encoder.
+
+    """
+
+    def __init__(
+        self,
+        idim: int = 80,
+        gst_tokens: int = 10,
+        gst_token_dim: int = 256,
+        gst_heads: int = 4,
+        conv_layers: int = 6,
+        conv_chans_list: List[int] = [32, 32, 64, 64, 128, 128],
+        conv_kernel_size: int = 3,
+        conv_stride: int = 2,
+        gru_layers: int = 1,
+        gru_units: int = 128,
+    ):
+        """Initilize global style encoder module."""
+        super(StyleEncoder, self).__init__()
+        self.ref_enc = ReferenceEncoder(
+            idim=idim,
+            conv_layers=conv_layers,
+            conv_chans_list=conv_chans_list,
+            conv_kernel_size=conv_kernel_size,
+            conv_stride=conv_stride,
+            gru_layers=gru_layers,
+            gru_units=gru_units,
+        )
+        self.stl = StyleTokenLayer(
+            ref_embed_dim=gru_units,
+            gst_tokens=gst_tokens,
+            gst_token_dim=gst_token_dim,
+            gst_heads=gst_heads,
+        )
+
+    def forward(self, speech: torch.Tensor) -> torch.Tensor:
+        """Calculate forward propagation.
+
+        Args:
+            speech (Tensor): Batch of padded target features (B, Lmax, odim).
+
+        Returns:
+            Tensor: Style token embeddings (B, token_dim).
+
+        """
+        ref_embs = self.ref_enc(speech)
+        style_embs = self.stl(ref_embs)
+
+        return style_embs
+
+
+class ReferenceEncoder(torch.nn.Module):
+    """Reference encoder module.
+
+    This module is refernece encoder introduced in `Style Tokens: Unsupervised Style
+    Modeling, Control and Transfer in End-to-End Speech Synthesis`.
+
+    .. _`Style Tokens: Unsupervised Style Modeling, Control and Transfer in End-to-End
+        Speech Synthesis`: https://arxiv.org/abs/1803.09017
+
+    Args:
+        idim (int, optional): Dimension of the input mel-spectrogram.
+        conv_layers (int, optional): The number of conv layers in the reference encoder.
+        conv_chans_list: (List[int], optional):
+            List of the number of channels of conv layers in the referece encoder.
+        conv_kernel_size (int, optional):
+            Kernal size of conv layers in the reference encoder.
+        conv_stride (int, optional):
+            Stride size of conv layers in the reference encoder.
+        gru_layers (int, optional): The number of GRU layers in the reference encoder.
+        gru_units (int, optional): The number of GRU units in the reference encoder.
+
+    """
+
+    def __init__(
+        self,
+        idim=80,
+        conv_layers: int = 6,
+        conv_chans_list: List[int] = [32, 32, 64, 64, 128, 128],
```

Comment:
The mutable default arg is stateful and gonna be bug. How about tuple or None instead? https://docs.python-guide.org/writing/gotchas/

Reply:
Thanks. Chaneged to Tuple.

Reply:
@ShigekiKarita `Union[Tuple, List] = (...)` is OK?